### PR TITLE
[WIP] feature(actions): actions cache, take 2

### PR DESCRIPTION
**Pull request by daniellacosse**
(TODO: fill out description)

### CODE REVIEW

File: scripts/run_action.mjs [hunk 1]
```diff
@@ -14,78 +14,167 @@
...
+      childProcess.on('close', code => {
+        if (code === 0) {
+          resolve(childProcess);
+        } else {
+          reject(childProcess);
+        }
+      });
+    });
+
+  let resolvedPath;
```
Comment (daniellacosse) [inline scripts/run_action.mjs:108]: i inlined a lot of these one-off functions, like you like

File: src/electron/build.action.sh [hunk 2]
```diff
@@ -49,7 +49,7 @@ if [[ -n ${SENTRY_DSN:-} ]]; then
     readonly SENTRY_URL="https://sentry.io/api/$PROJECT_ID/store/?sentry_version=7&sentry_key=$API_KEY"
 fi
 
-readonly WEBPACK_MODE="$(node ./scripts/get_webpack_mode.mjs --buildMode=${BUILD_MODE})"
+readonly WEBPACK_MODE="$(node ./scripts/get_webpack_build_mode.mjs --buildMode=${BUILD_MODE})"
```
Comment (daniellacosse) [inline src/electron/build.action.sh:52]: oops 😅

File: scripts/run_action.mjs [hunk 3]
```diff
@@ -14,87 +14,169 @@
...
+  const pathRoots = [process.env.ROOT_DIR, path.join(process.env.ROOT_DIR, 'src')];
```
Comment (fortuna) [inline scripts/run_action.mjs:109]: Can you move these to some `resolveActionPath` function?
It will help the readability of `runAction`

File: scripts/run_action.mjs [hunk 4]
```diff
@@ -14,87 +14,169 @@
...
+
+  if (inputs.length) {
+    for (const input of inputs) {
+      if (!existsSync(input)) {
+        throw new ReferenceError(`action(${actionPath}) requires the input: "${input}". Aborting!`);
+      }
+    }
+
+    let mostRecentlyModifiedInput = 0;
+    let inputsToCheck = [...inputs];
```
Comment (fortuna) [inline scripts/run_action.mjs:141]: This only considers direct inputs.

If you have f1 -> action1 -> f2 -> action2 -> f3

If f1 changes and you run action2, you will miss the update to f1.

You need to navigate the entire tree of dependencies. You need to check not only the direct dependencies, but also the dependencies of the actions you call. It's a two-pass approach.

I guess the logic is:
1. (Always) Run all your direct dependencies
2. If any direct dependency got updated, run the action

Build systems are not easy. I wonder if it makes sense to use NX or bazel instead, or even make.
Perhaps it's worth playing with what they would look like at some point, before investing too much on our own build system.

**Review by fortuna — CHANGES_REQUESTED** [review]
I believe this is not handling transitive dependencies correctly.

File: scripts/run_action.mjs [hunk 5]
```diff
@@ -14,87 +14,169 @@
...
+ * @param {string[]} options.inputs Files and folders the action consumes. Specify these to skip the action when able
  * @returns {void}
  */
-export async function runAction(actionPath, ...parameters) {
-  const resolvedPath = await resolveActionPath(actionPath);
+export async function runAction(actionPath, {parameters = [], inputs = []} = {}) {
+  /**
+   * @description promisifies the child process spawner
+   */
+  const spawnPromise = (command, parameters) =>
```
[the remark this answers is not in the document]
Comment (daniellacosse) [reply scripts/run_action.mjs:84]: I was moving it in like you wanted me to do here: https://github.com/Jigsaw-Code/outline-client/pull/1264#discussion_r861353396

I'd prefer to be consistent, so either we move that out or keep this in

File: scripts/run_action.mjs (hunk 4 above)
Comment (daniellacosse) [reply scripts/run_action.mjs:141]: I don't understand how this doesn't do that! In this PR, each action when called is checking its own dependencies. So we effectively navigate the entire tree of dependencies manually and recursively! For instance:

https://github.com/Jigsaw-Code/outline-client/pull/1291/files#diff-3ade035b399e6b2f6083fc0cbee893d3e5683f71af90d77c5e7a70837ca5bd23R31-R32

The first `runAction` call on line 31 may create a new `www`. This `www` is explicitly passed into the input call in the second action call of `runAction` on line 32. So that call will see if that folder changes and be called anyway. And this all is tracked within the top level action we're in - it makes it very clear (to me, perhaps it needs some tweaking) what depends on which when I read the action file and what the chain of dependencies are.

File: scripts/run_action.mjs [hunk 6]
```diff
@@ -14,87 +14,169 @@
...
+    while (inputsToCheck.length) {
+      const currentFolderOrFile = inputsToCheck.pop();
+      const fileInformation = await fs.stat(currentFolderOrFile);
+
+      mostRecentlyModifiedInput = Math.max(mostRecentlyModifiedInput, fileInformation.mtimeMs);
+
+      if (fileInformation.isDirectory()) {
+        const contents = await fs.readdir(currentFolderOrFile);
+
+        inputsToCheck = [...inputsToCheck, ...contents.map(child => `${currentFolderOrFile}/${child}`)];
```
[the remark this answers is not in the document]
Comment (daniellacosse) [reply scripts/run_action.mjs:151]: I was just being functional here. Force of habit.

File: scripts/run_action.mjs [hunk 7]
```diff
@@ -14,78 +14,167 @@
...
+ * Cache methods:
+ *   readActionCache - loads the latest time the action was run along with the parameters it was run with
+ *   writeActionCache - updates the time an action was run and the parameters it was run with
+ *
+ * TODO(daniellacosse): convert the cache to a folder so each actions can run in parallel without risking
+ * corruption of the cache
  */
-const resolveActionPath = async actionPath => {
-  if (!actionPath) return '';
+class ActionCache {
```
[the remark this answers is not in the document]
Comment (daniellacosse) [reply scripts/run_action.mjs:35]: I'm not sure about calling it `build` since that could get confusing - we have `build` actions in all our other folders, too. I'll think about this.

File: .gitignore [hunk 8]
```diff
@@ -1,3 +1,4 @@
+.action_cache.json
```
[the remark this answers is not in the document]
Comment (daniellacosse) [reply .gitignore:1]: Or whatever we call it, sure.

File: scripts/run_action.mjs [hunk 9]
```diff
@@ -14,87 +14,169 @@
...
+  ACTION_CACHE_FILE = './.action_cache.json';
```
[the remark this answers is not in the document]
Comment (daniellacosse) [reply scripts/run_action.mjs:36]: If it did the CI would fail.

File: scripts/run_action.mjs (hunk 4 above)
Comment (fortuna) [reply scripts/run_action.mjs]: As discussed in our meeting, it's important to make sure the build is correct.
Even if it's correct with the existing actions, changes and new actions can break without warning. That has caused us significant pain in the past.

I believe this can be fixed by checking the transitive inputs first.
Alternatively, instead of having an actions cache in the build system, specific actions could be made incremental. Similar to `cordova prepare`, but taking into account the inputs properly.

Comment (daniellacosse) [reply scripts/run_action.mjs]: Actually, thinking about it over the weekend - what if we stopped thinking about this as a build system and more as a "scripting framework"?

The reason this currently works, and why the `'cordova/setup'` call here -

```js
  await runAction('www/build', {parameters: [`--buildMode=${buildMode}`], inputs: ['src/www']});
  await runAction('cordova/setup', {parameters, inputs: ['www', 'resources', 'src/cordova', 'cordova-plugin-outline']});
```

- doesn't have to also include `"src/www"` as an input is because [`cordova/setup` itself calls the `"www/build"` action as well with `"src/www"` as input](https://github.com/Jigsaw-Code/outline-client/pull/1291/files#diff-83e4918643471872d89a312fdd8603e164f6bc00c21c6bce3c877b07821898e0R52). Basically the way I have it set up now, if each action calls _all_ its child actions directly, it works. Yes, this results in some redundant code, but it's no better than what we had before. In fact we could have each action declare "input defaults" and then we wouldn't even need to think about what inputs are needed anymore. I can even add an option `skipIfNoInputChanges` to make it opt-in and more clear what's happening when you read the file.

I actually like this "each action calls all its child actions" approach because it allows us to specify when we write a new action how its dependencies should be called. Imagine an action that builds all platforms - if we statically declare a list of actions we need to call, we lose flexibility of potentially being able to build and rmraf `www` between platform builds. I don't think Nx could give us something like that.

That said, I'm not against refactoring this system just that each action DFSes its child components to determine skipability, however - turning actions into extensible classes would require typescript interfaces, which would require moving everything to `ts-node` - and at that point I'm thinking maybe the work is similar in scope to trying to add a bunch of `package.json`s to our `src` directory and leveraging workspace tools?

Basically, is "each action must explicitly call all its child actions" sufficient enough of a rule that we can get the time savings of this PR for now and refactor things into our ideal state a bit later?
